src/tuneflow_py/models/lyric.py

```diff
@@ -0,0 +1,420 @@
+                tokens.append(char)
+            elif unicodedata.category(char).startswith('P'):
+                # Handle punctuation marks
+                if current_token:
+                    if re.match(r"^[']+$", char) and re.match(r'^[A-Za-z0-9]+$', current_token):
+                        current_token += char
+                    else:
+                        tokens.append(current_token)
+                        tokens.append(char)
+                        current_token = ''
+                else:
+                    tokens.append(char)
+            else:
+                current_token += char
+
+        if current_token:
+            tokens.append(current_token)
+
+        return tokens
+
+
+class Lyrics:
+    def __init__(self, song: Song) -> None:
+        self.song = song
+        if song._proto.lyrics is None:
+            song._proto.lyrics = song_pb2.Lyrics()
+        self._proto = song._proto.lyrics
+
+    def create_line(self, start_tick: int, resolve_order: bool = True):
+        '''
+        Creates a new empty LyricLine object at the given start_tick.
+
+        Args:
+            start_tick (int): The starting tick for the new lyric line.
+            resolve_order (bool, optional): Whether to sort the lyric lines by start_tick. Defaults to True.
+
+        Returns:
+            LyricLine: The newly created LyricLine object.
+        '''
+        line_proto = self._proto.lines.add()
+        line = LyricLine(lyrics=self, start_tick=start_tick, proto=line_proto)
+        if resolve_order:
+            self.sort_lines()
+        return line
+
+    def create_line_from_string(self, input: str, start_tick: int, end_tick: int, tokenizer: Optional[LyricTokenizer] = None):
+        """
+        Create a Line object from a string of words.
+
+        Args:
+            input (str): The input string of words.
+            start_tick (int): The starting tick of the line.
+            end_tick (int): The ending tick of the line.
+            tokenizer (Optional[LyricTokenizer], optional): The tokenizer to use for splitting the input string into words.
+                If None, the default tokenizer will be used. Defaults to None.
+
+        Returns:
+            Line: The created Line object.
+
+        Raises:
+            ValueError: If the input string is empty.
+        """
+        line_proto = self._proto.lines.add()
+        line = LyricLine(lyrics=self, start_tick=start_tick, proto=line_proto)
+        if not input:
+            return line
+        line.create_words_from_string(input, start_tick, end_tick, tokenizer)
+        return line
+
+    @staticmethod
+    def _get_index_of_line_at_tick(proto: song_pb2.Lyrics, tick: int):
+        '''
+        Internal generator implementation that locates lyrlic line at a given tick.
+        '''
+        if len(proto.lines) == 0:
+            return -1
+        start_search_index = lower_equal(
+            [LyricLine._get_start_tick(line) for line in proto.lines],
```

Comment:
This invalidates the whole purpose of this binary search since its complexity is already O(N). You could pass a key function that checks for the type of element and returns tick if it is a number, or LyricLine._get_start_tick(line) otherwise.